### PR TITLE
Add ReadIterator for Query & Scan

### PR DESCRIPTION
**Pull request by borgstrom**
**Breaking change**

This PR introduces the `ReadIterator` for Query & Scan operations.

This is a [iterator object](https://docs.python.org/3/library/stdtypes.html#typeiter) that allows us some sugar when working with requests & results.

Previously calls to query or scan would recursively execute, with no way to opt out.  Experience over the past ~year has led us to believe that is undesirable, and that in-fact recursion should be fully opt-in so that users can more efficiently control requests to DynamoDB.

Now, the iterator object exposes a `.last` attribute that will be set to the primary key of the last object in the result set and a `.start()` helper.  This can be passed to clients so that on subsequent calls they can provide it to resume the results from that point.

```
books = Books.scan()
last = get_last_from_client()
if last:
    books.start(last)
return {'items': list(books), 'last': books.last}
```

The iterator object also provides a method to set a limit for queries.

```
books = Books.scan().limit(5)
```

And one to enable recursion, which mimics the old behavior.

```
books = Books.scan().recursive()
```

### CODE REVIEW

**Comment by borgstrom** [comment]
FYI @andyinsf @dpnnw

File: dynamorm/table.py [hunk 1]
```diff
@@ -773,3 +773,178 @@ def Q(**mapping):
...
+                pename = '#pe{}'.format('_'.join([str(attri), str(parti)]))
+                self.dynamo_kwargs['ExpressionAttributeNames'][pename] = part
+                name_parts.append(pename)
+            pe.append('.'.join(name_parts))
+
+        self.dynamo_kwargs['ProjectionExpression'] = ', '.join(pe)
+        self._partial = True
+        return self
+
+    def recursive(self):
```
Comment (ahollenbach) [inline dynamorm/table.py]: I'm not quite following what makes the continuous version recursive - isn't it more of an unbounded iterator?

**Comment by ahollenbach** [comment]
Did a high-level pass and it looks really good. I can dig deeper and poke at the implementation if nobody else will be looking

File: dynamorm/table.py (hunk 1 above)
Comment (borgstrom) [reply dynamorm/table.py]: What’s the difference between recursion and unbounded iteration?

Comment (borgstrom) [reply dynamorm/table.py]: Rather, what would make more sense to you for a method name/attr?

**Comment by borgstrom** [comment]
Merging.  We can address any feedback anyone has in followup PRs if needed.

File: dynamorm/table.py (hunk 1 above)
Comment (ahollenbach) [reply dynamorm/table.py]: OH. I missed __next__ calling itself

Comment (ahollenbach) [reply dynamorm/table.py]: I am having trouble reading through this iterator code - I'll have to take a look when I'm feeling more fresh or something

**Pull request merged**